libs/s25main/mapGenerator/DistanceByProperty.h

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2005 - 2017 Settlers Freaks (sf-team at siedler25.org)
+//
+// This file is part of Return To The Roots.
+//
+// Return To The Roots is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Return To The Roots is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef DistanceField_h__
+#define DistanceField_h__
+
+#include "mapGenerator/Map.h"
+#include "mapGenerator/GridUtility.h"
+#include "gameTypes/MapCoordinates.h"
+
+#include <vector>
+#include <queue>
+#include <set>
+#include <map>
+#include <cmath>
+
+namespace rttr {
+namespace mapGenerator {
+
+/**
+ * Evalutor to detect non-buildable positions on a map.
+ * @param map reference to the actual map
+ * @param index index of the position to check
+ * @return 'true' if the position with the specified index is not buildable, 'false' otherwise.
+ */
+bool IsNotBuildable(const Map& map, int index);
+
+/**
+ * Evalutor to detect head quarter positions on a map.
+ * @param map reference to the actual map
+ * @param index index of the position to check
+ * @return 'true' if the position with the specified index is a HQ, 'false' otherwise.
+ */
+bool IsHeadQuarter(const Map& map, int index);
+
+/**
+ * Evalutor to detect non-water area on a map.
+ * @param map reference to the actual map
+ * @param index index of the position to check
+ * @return 'true' if the position with the specified index is not water, 'false' otherwise.
+ */
+bool IsLand (const Map& map, int index);
+
+/**
+ * Evalutor to detect coast area on a map.
+ * @param map reference to the actual map
+ * @param index index of the position to check
+ * @return 'true' if the position with the specified index is part of a coast, 'false' otherwise.
+ */
+bool IsCoastLand(const Map& map, int index);
+
+/**
+ * Computes a vector field describing the distance of each grid position (by index) to the closest position
+ * for which the evaluator returned true.
+ * @param args generic parameter which is passed to the evaluator along with the index of the grid index
+ * @param size absolute size of the current map in x- and y-direction
+ * @param evaluator function used evaluate whether or not the property is given
+ * @return distance of each grid position to closest point with the evaluated property being true.
+ */
+template<typename T>
+std::vector<int> DistanceByProperty(const T& args,
```

Comment:
Suggestion for better design (IMO): Use `template<typename T> std::vector<int> DistanceByProperty(const MapExtent& size, T&& evaluator)`. This avoids the "old-shool" callback-with-param based approach and you can use lambdas instead as well as raw function pointers. Downside is that you need to document the call syntax. Or you can use `std::function<bool(int)>` and get rid of the template but pay for a bit of runtime cost due to a virtual call, but branch prediction should rescue some.

Reply:
**Fixed**
Good point, I updated the `DistanceByProperty ` function.